src/main/java/org/commcare/formplayer/configuration/WebSecurityConfig.java

```diff
@@ -0,0 +1,105 @@
+import org.springframework.security.web.firewall.StrictHttpFirewall;
+
+@Configuration
+@EnableWebSecurity
+public class WebSecurityConfig extends WebSecurityConfigurerAdapter {
+    @Autowired
+    private HqUserDetailsService userDetailsService;
+
+    @Autowired
+    private FormSessionService formSessionService;
+
+    @Value("${commcarehq.formplayerAuthKey}")
+    private String formplayerAuthKey;
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        disableDefaults(http);
+
+        http
+            .authorizeRequests()
+            .antMatchers("/serverup", "/validate_form", "/favicon.ico")
+            .permitAll();
+        http.authorizeRequests().antMatchers("/**").authenticated();
+        http.addFilterAt(getHmacAuthFilter(), UsernamePasswordAuthenticationFilter.class);
+        http.addFilterAfter(sessionAuthFilter(), HmacAuthFilter.class);
+        http.cors();
+    }
+
+    private HmacAuthFilter getHmacAuthFilter() {
+        return HmacAuthFilter.builder()
+                .hmacKey(formplayerAuthKey)
+                .formSessionService(formSessionService)
+                .build();
+    }
+
+    private void disableDefaults(HttpSecurity http) throws Exception {
+        http.csrf().disable();
+        http.sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS);
+        http.requestCache().disable();  // only needed for login workflow
+        http.logout().disable();
+        http.formLogin().disable();
+        http.httpBasic().disable();
+    }
+
+    public CommCareSessionAuthFilter sessionAuthFilter() throws Exception {
+        CommCareSessionAuthFilter filter = new CommCareSessionAuthFilter();
+        filter.setAuthenticationManager(authenticationManagerBean());
+        filter.setRequiresAuthenticationRequestMatcher(new CommCareSessionAuthFilter.SessionAuthRequestMatcher());
+        return filter;
+    }
+
+    /**
+     * Configure the authentication manager with a provider that will use the
+     * ``HQUserDetailsService`` to authenticate the user.
+     */
+    @Override
+    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
+        PreAuthenticatedAuthenticationProvider provider = new PreAuthenticatedAuthenticationProvider();
+        provider.setPreAuthenticatedUserDetailsService(userDetailsService);
+        auth.authenticationProvider(provider);
+    }
+
+    @Override
+    public void configure(WebSecurity web) throws Exception {
+        web.httpFirewall(allowDoubleSlashHttpFirewall());
+    }
+
+    /**
+     * Temporary customization of the firewall to allow '//' in the URL.
+     * This should be removed once all 3rd party hosters have been given an opportunity
+     * to update their proxy configuration.
+     */
+    @Bean
+    public HttpFirewall allowDoubleSlashHttpFirewall() {
+        StrictHttpFirewall firewall = new StrictHttpFirewall();
+        firewall.setAllowUrlEncodedDoubleSlash(true);
```

Comment:
I've added this so that we can decouple this PR from the proxy changes in https://github.com/dimagi/commcare-cloud/pull/4584